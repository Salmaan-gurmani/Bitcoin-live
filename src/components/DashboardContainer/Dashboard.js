import React, { useEffect, useState } from 'react';

import {
    AppBar,
    Box,
    Toolbar,
    Button,
    Typography,
    Paper,
    Grid,
} from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import * as R from 'ramda';
import { loadCurrentPrice, loadPriceHistory } from '../../actions';
import { useStyles } from './Dashboard.styles';
import GraphComponent from '../GraphComponent/GraphComponent';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import { useInterval } from './helper';

const HomePage = () => {
    const bitCoinPrice = useSelector(state => state.bitcoinPrice);
    const [currency, setCurrency] = useState('USD');
    const error = useSelector(state => state.error);
    const bitCoinPriceHistory = useSelector(state => state.bitcoinPriceHistory);
    const classes = useStyles();
    const dispatch = useDispatch();
    const hasCurrentPrice = R.hasPath(
        ['response', 'bpi', currency, 'rate'],
        bitCoinPrice,
    );
    const hasPriceHistory = R.hasPath(['response', 'bpi'], bitCoinPriceHistory);

    useEffect(() => {
        dispatch(loadCurrentPrice({ currency: currency }));
        dispatch(loadPriceHistory({ currency: currency }));
    }, [currency]);

    const handleChange = event => {
        setCurrency(event.target.value);
    };

    useInterval(() => {
        dispatch(loadCurrentPrice({ currency: currency }));
    }, 1000);

    return (
        <div>
            <Box sx={{ flexGrow: 1 }}>
                <AppBar position="static" className={classes.bar}>
                    <Toolbar>
                        <img
                            src="https://logos-download.com/wp-content/uploads/2016/09/Adecco_logo.png"
                            alt="logo"
                            className={classes.icon}
                        />
                        <Button className={classes.loginButton} color="inherit">
                            Login
                        </Button>
                    </Toolbar>
                </AppBar>
            </Box>
            <div className={classes.root}>
                <FormControl variant="outlined" className={classes.formControl}>
                    <InputLabel id="demo-simple-select-outlined-label">
                        Currency
                    </InputLabel>
                    <Select
                        labelId="demo-simple-select-outlined-label"
                        id="demo-simple-select-outlined"
                        value={currency}
                        onChange={handleChange}
                        label="Currency"
                    >
                        <MenuItem value={'USD'}>USD</MenuItem>
                        <MenuItem value={'EUR'}>EUR</MenuItem>
                        <MenuItem value={'CNY'}>CNY</MenuItem>
                        <MenuItem value={'JPY'}>JPY</MenuItem>
                        <MenuItem value={'PLN'}>PLN</MenuItem>
                    </Select>
                </FormControl>
                <div className={classes.secondaryContainer}>
                    {hasCurrentPrice && (
                        <Typography variant="h4" color="primary">
                            Current Price in {currency}:{' '}
                        </Typography>
                    )}
                    {hasCurrentPrice && (
                        <Typography variant="h4" color="primary">
                            {bitCoinPrice.response.bpi[currency].rate}
                        </Typography>
                    )}
                </div>

                {hasPriceHistory && (
                    <Grid container>
                        <Grid item xs={12}>
                            <Paper>
                                <GraphComponent
                                    priceHistory={
                                        bitCoinPriceHistory.response.bpi
                                    }
                                    currency={currency}
                                />
                            </Paper>
                        </Grid>
                    </Grid>
                )}
            </div>

            {error && (
                <div className={classes.error}>{JSON.stringify(error)}</div>
            )}
        </div>
    );
};

export default HomePage;
