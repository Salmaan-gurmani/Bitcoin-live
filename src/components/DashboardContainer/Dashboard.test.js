import * as React from 'react';
import Dashboard from './Dashboard';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import * as redux from 'react-redux';

configure({ adapter: new Adapter() });

const bitcoinPriceData = {
    response: {
        time: {
            updated: 'Nov 23, 2021 21:41:00 UTC',
            updatedISO: '2021-11-23T21:41:00+00:00',
            updateduk: 'Nov 23, 2021 at 21:41 GMT',
        },
        disclaimer:
            'This data was produced from the CoinDesk Bitcoin Price Index (USD). Non-USD currency data converted using hourly conversion rate from openexchangerates.org',
        bpi: {
            USD: {
                code: 'USD',
                rate: '57,601.2737',
                description: 'United States Dollar',
                rate_float: 57601.2737,
            },
        },
    },
};

const bitCoinPriceHistoryData = {
    response: {
        bpi: {
            '2021-10-23': 61307.6667,
            '2021-10-24': 60849.5667,
            '2021-10-25': 63084.5717,
            '2021-10-26': 60307.6533,
            '2021-10-27': 58457.8717,
            '2021-10-28': 60591.9433,
            '2021-10-29': 62253.32,
            '2021-10-30': 61888.405,
            '2021-10-31': 61338.5317,
            '2021-11-01': 60955.7683,
            '2021-11-02': 63246.855,
            '2021-11-03': 62926.8233,
            '2021-11-04': 61444.2283,
            '2021-11-05': 61016.2883,
            '2021-11-06': 61514.3833,
            '2021-11-07': 63287.79,
            '2021-11-08': 67544.8733,
            '2021-11-09': 66941.8717,
            '2021-11-10': 64924.3533,
            '2021-11-11': 64818.75,
            '2021-11-12': 64142.9817,
            '2021-11-13': 64410.975,
            '2021-11-14': 65495.3183,
            '2021-11-15': 63599.7467,
            '2021-11-16': 60112.9583,
            '2021-11-17': 60364.085,
            '2021-11-18': 56925.4283,
            '2021-11-19': 58121.7817,
            '2021-11-20': 59763.9583,
            '2021-11-21': 58689.5383,
            '2021-11-22': 56290.1683,
        },
        disclaimer:
            'This data was produced from the CoinDesk Bitcoin Price Index. BPI value data returned as USD.',
        time: {
            updated: 'Nov 23, 2021 00:03:00 UTC',
            updatedISO: '2021-11-23T00:03:00+00:00',
        },
    },
};
describe('<Dashboard />', () => {
    let useEffect;
    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };
    const useSelectorMock = jest.spyOn(redux, 'useSelector');
    const useDispatchMock = jest.spyOn(redux, 'useDispatch');

    beforeEach(() => {
        useSelectorMock.mockClear();
        useDispatchMock.mockClear();
        useEffect = jest.spyOn(React, 'useEffect');
        mockUseEffect(); // 2 times
        mockUseEffect(); //
    });
    it('should render Dashboard', () => {
        useSelectorMock.mockImplementation(func =>
            func({
                bitcoinPrice: bitcoinPriceData,
                bitCoinPriceHistory: bitCoinPriceHistoryData,
            }),
        );
        const wrapper = shallow(<Dashboard />);
        expect(wrapper).toMatchSnapshot();
    });
    it('should render Dashboard with error', () => {
        useSelectorMock.mockImplementation(func =>
            func({
                bitcoinPrice: bitcoinPriceData,
                bitCoinPriceHistory: bitCoinPriceHistoryData,
                error: true,
            }),
        );

        const wrapper = shallow(<Dashboard />);
        expect(wrapper).toMatchSnapshot();
    });
});
