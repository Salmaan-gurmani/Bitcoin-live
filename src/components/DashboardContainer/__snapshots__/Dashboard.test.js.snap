// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> should render Dashboard 1`] = `ShallowWrapper {}`;

exports[`<Dashboard /> should render Dashboard with error 1`] = `ShallowWrapper {}`;
